content/client-lib-development-guide/features.textile

```diff
@@ -237,6 +237,7 @@ h3(#rest-channel). Channel
 *** @(RSL1k3)@ If more than one @Message@ is passed to @publish()@ and one or more of those messages contains a non-empty @id@ attribute, then all message ids (present or absent) are preserved on sending the batch of messages.
 *** @(RSL1k4)@ An explicit test for idempotency of publishes with library-generated ids shall exist that simulates an error response to a successful publish of a batch of messages, expects an automatic retry by the library, and verifies that the net outcome is that the batch is published only once.
 *** @(RSL1k5)@ An explicit test for idempotency of publishes with client-supplied ids shall exist that involves multiple explicit publish requests for a given message and verifies that the net outcome is that the message is published only once.
+** @(RSL1l)@ The @publish(Message)@ and @publish(Message[])@ forms of the method should take an extra @Dict<String, Stringifiable>@ argument. These parameters should be encoded using normal querystring-encoding and sent as part of the query string of the REST publish. (@Stringifiable@ is defined in @RTC1f@)
```

Comment:
Can we suggest a test here please so that a developer knows the params are workign?  Otherwise it's hard to know if the implementation works as expected for every dev who implements this feature without us checking it.

Reply:
You mean simply state that a test must exist that proves the params are included the request (eg by intercepting the request the library makes, a mock http layer or echo server); or suggest a test that actually passes the params to realtime, so that the effect of that is verifiable in the result?

Reply:
> or suggest a test that actually passes the params to realtime, so that the effect of that is verifiable in the result?

The latter, that is what we have always aimed to do with SDK tests no? We build our SDKs to test against sandbox for a reason, and stubbing out incorrect params or a bad implementation doesn't prove the SDK does, it just gives the developer a happy feeling.

Reply:
How about an undocumented `_forceNack=true` param that forces the publish to be rejected with a particular error code, that the client lib developer can check for?

Reply:
^ https://github.com/ably/docs/pull/742/commits/eec87554204df72f846c474f467d598224c39392